src/Stratis.Bitcoin/FinalizedBlockInfoRepository.cs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    public class FinalizedBlockInfoRepository : IFinalizedBlockInfoRepository
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>Database key under which the block height of the last finalized block height is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height and hash of a block that can't be reorged away from.</summary>
+        private HashHeightPair finalizedBlockInfo;
+
+        /// <summary>Queue of finalized infos to save.</summary>
+        /// <remarks>All access should be protected by <see cref="queueLock"/></remarks>
+        private readonly Queue<HashHeightPair> finalizedBlockInfosToSave;
+
+        /// <summary>Protects access to <see cref="finalizedBlockInfosToSave"/>.</summary>
+        private readonly object queueLock;
+
+        /// <summary>Task that continously persists finalized block info to the database.</summary>
+        private readonly Task finalizedBlockInfoPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        private readonly AsyncManualResetEvent queueUpdatedEvent;
+
+        public FinalizedBlockInfoRepository(string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.dbreeze = new DBreezeEngine(folder);
+
+            this.finalizedBlockInfosToSave = new Queue<HashHeightPair>();
+            this.queueLock = new object();
+
+            this.queueUpdatedEvent = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+            this.finalizedBlockInfoPersistingTask = this.PersistFinalizedBlockInfoContinuouslyAsync();
+        }
+
+        public FinalizedBlockInfoRepository(DataFolder dataFolder, ILoggerFactory loggerFactory)
+            : this(dataFolder.FinalizedBlockInfoPath, loggerFactory)
+        {
+        }
+
+        private async Task PersistFinalizedBlockInfoContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.queueUpdatedEvent.WaitAsync(this.cancellation.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                }
+
+                this.queueUpdatedEvent.Reset();
+
+                HashHeightPair lastFinalizedBlock = null;
+
+                lock (this.queueLock)
+                {
+                    while (this.finalizedBlockInfosToSave.Count != 0)
```

Comment:
in the queue we will have: 

456, 457, 458, 459.

Only  459 should be saved. 

So the logic is correct and intended. It's there in case we are pushing new finalized block infos faster than we can persist them on the harddrive. When this happens we ONLY need the most recent one.

Reply:
What if the node crashes? how do we make sure finalized is in sync with consensus?

Reply:
If the node crashes we will have finalized height that was saved 100ms before the crush. 
And we will update it with the first block being connected so it's all good. 
Plus on startup we recalculate the actual finalized block hash (because we rewind consensus in case of crush and finalized block info with it)

Reply:
"So the logic is correct and intended. It's there in case we are pushing new finalized block infos faster than we can persist them on the harddrive. When this happens we ONLY need the most recent one."

OK 💯 makes perfect sense. But perhaps add a comment on top to explain the logic? Unless I'm just being stupid here ;)

Reply:
Fixed